src/common/metrics.hpp

```diff
@@ -18,21 +20,31 @@ class Metrics;
 
 using MetricsPtr = std::shared_ptr<Metrics>;
 
-
 class Metrics {
 public:
 
   Metrics(Logger logger);
 
-  Counter add(const std::string& name, const std::string& help, const std::string& output_type) noexcept;
+  template<typename MetricType, typename... Args>
+  MetricType add(MetricsContext context, Args&&... metric_args);
+  
   void register_on(prometheus::Exposer& exposer);
 
 private:
 
-  using RegistryPtr = std::shared_ptr<prometheus::Registry> ;
+  using RegistryPtr = std::shared_ptr<prometheus::Registry>;
 
   Logger       m_logger;
   RegistryPtr  m_registry;
 };
 
+
+// Avaliable MetricTypes: Counter, Histogram
+// Args for Counter: None
+// Args for Histogram: std::vector<double> bounds
+template<typename MetricType, typename... Args>
+MetricType Metrics::add(MetricsContext context, Args &&... metric_args)  {
```

Comment:
I don't really like this design, for several reasons:
1. it __requires__ comments to understand how to use this API.
2. It is not clear how to create e.g. `Counter` just looking at `counter.hpp` (though this problem was before this PR)

The purpose of this function is to provide `m_registry` for `Metric` constructor, this could be done via private `registry()` getter, available only for Metric (Counter, Histogram) types and providing reference to `Metrics` object to constructor of metric. 

..ugh. `Metrics` is not a good name for this class, It should be called `Registry` and placed into `metrics` namespace with all metric types (`Description`, `Counter`, `Histogram`).